android/src/main/kotlin/com/rudderstack/sdk/kotlin/android/plugins/devicemode/IntegrationsManagementPlugin.kt

```diff
@@ -0,0 +1,173 @@
+
+        if (queuedEventResult.isFailure) {
+            dropOldestAndSend(event)
+        }
+
+        return event
+    }
+
+    override fun teardown() {
+        integrationPluginChain.removeAll()
+        queuedEventsChannel.cancel()
+        destinationReadyCallbacks.clear()
+    }
+
+    internal fun onDestinationReady(plugin: IntegrationPlugin, onReady: (Any?, DestinationResult) -> Unit) {
+        destinationReadyCallbacks
+            .getOrPut(plugin.key) { mutableListOf() }
+            .add(onReady)
+
+        if (integrationPluginChain.hasIntegration(plugin)) {
+            notifyOnDestinationReady(plugin)
+        }
+    }
+
+    internal fun addIntegration(plugin: IntegrationPlugin) {
+        integrationPluginChain.add(plugin)
+        if (isSourceEnabled) {
+            initAndNotifyReady(analytics.sourceConfigState.value, plugin)
+        }
+    }
+
+    internal fun removeIntegration(plugin: IntegrationPlugin) {
+        integrationPluginChain.remove(plugin)
+    }
+
+    internal fun reset() {
+        integrationPluginChain.applyClosure { plugin ->
+            if (plugin is IntegrationPlugin) {
+                if (plugin.destinationState.isReady()) {
+                    plugin.reset()
+                } else {
+                    LoggerAnalytics.debug(
+                        "IntegrationsManagementPlugin: Destination ${plugin.key} is " +
+                            "not ready yet. Reset discarded."
+                    )
+                }
+            }
+        }
+    }
+
+    internal fun flush() {
+        integrationPluginChain.applyClosure { plugin ->
+            if (plugin is IntegrationPlugin) {
+                if (plugin.destinationState.isReady()) {
+                    plugin.flush()
+                } else {
+                    LoggerAnalytics.debug(
+                        "IntegrationsManagementPlugin: Destination ${plugin.key} is " +
+                            "not ready yet. Flush discarded."
+                    )
+                }
+            }
+        }
+    }
+
+    private fun initAndNotifyReady(sourceConfig: SourceConfig, plugin: IntegrationPlugin) {
+        plugin.initialize(sourceConfig)
+        notifyOnDestinationReady(plugin)
+    }
+
+    @Synchronized
+    private fun notifyOnDestinationReady(plugin: IntegrationPlugin) {
+        val callBacks = destinationReadyCallbacks[plugin.key]?.toList()
+        callBacks?.forEach { callback ->
+            when (val destinationState = plugin.destinationState) {
+                DestinationState.Ready -> notifyAndRemoveCallback(plugin, callback, Result.Success(Unit))
+                is DestinationState.Failed -> notifyAndRemoveCallback(
+                    plugin,
+                    callback,
+                    Result.Failure(error = destinationState.exception)
+                )
+                DestinationState.Uninitialised -> Unit
+            }
+        }
+    }
```

Comment:
We could improve the method name and a bit of code.

Regarding method name: It's just notifying the destination callbacks. Hence the term`OnDestinationReady` doesn't seem to explain the actual use case. I think we could improve it by renaming it to the `notifyDestinationCallbacks`.


```suggestion
    private fun notifyDestinationCallbacks(plugin: IntegrationPlugin) {
        val callBacks = destinationReadyCallbacks[plugin.key]?.toList()
        callBacks?.forEach { callback ->
            when (val destinationState = plugin.destinationState) {
                is DestinationState.Ready -> notifyAndRemoveCallback(plugin, callback, Result.Success(Unit))
                is DestinationState.Failed -> notifyAndRemoveCallback(
                    plugin,
                    callback,
                    Result.Failure(error = destinationState.exception)
                )
                is DestinationState.Uninitialised -> Unit
            }
        }
    }
```